lib/Transforms/IPO/StructFieldAccessInfo.cpp

```diff
@@ -0,0 +1,343 @@
+        addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse,
+                 Inst->getOpcode());
+      }
+    } else {
+      addStats(StructFieldAccessManager::DebugStats::
+                   DS_UserNotInstructionNorOperator);
+    }
+  }
+}
+
+void StructFieldAccessInfo::analyzeUsersOfStructValue(const Value *V) {
+  assert(V->getType()->isPointerTy() &&
+         V->getType()->getPointerElementType()->isStructTy());
+  for (auto *U : V->users()) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                    dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    if (isa<Instruction>(U)) {
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (!isa<GetElementPtrInst>(Inst)) {
+        // Only support access struct through GEP for now
+        if (Inst->getOpcode() == Instruction::Call) {
+          DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs()
+                                             << "User is a call instruction\n");
+          assert(Inst && isa<CallInst>(Inst));
+          auto *F = dyn_cast<CallInst>(Inst)->getCalledFunction();
+          if (!F || F->isDeclaration()) {
+            // If a struct is passed to an indirect or a function not declared
+            // in the program, we can't analyze it...
+            Eligiblity = false;
+          }
+        } else if (Inst->getOpcode() == Instruction::Invoke) {
+          DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                          dbgs() << "User is an invoke instruction\n");
+          assert(Inst && isa<InvokeInst>(Inst));
+          auto *F = dyn_cast<InvokeInst>(Inst)->getCalledFunction();
+          if (!F || F->isDeclaration()) {
+            // If a struct is passed to an indirect or a function not declared
+            // in the program, we can't analyze it...
+            Eligiblity = false;
+          }
+        }
+        continue;
+      }
+      addFieldAccessFromGEP(Inst);
+    } else if (isa<Operator>(U)) {
+      auto *Inst = dyn_cast<Operator>(U);
+      if (!isa<GEPOperator>(Inst)) {
+        // Only support access struct through GEP for now
+        continue;
+      }
+      addFieldAccessFromGEP(Inst);
+    } else {
+      addStats(StructFieldAccessManager::DebugStats::
+                   DS_UserNotInstructionNorOperator);
+    }
+  }
+}
+
+void StructFieldAccessInfo::analyzeUsersOfStructPointerValue(const Value *V) {
+  // Analyze users of value defined as struct*
+  assert(V->getType()->isPointerTy() &&
+         V->getType()->getPointerElementType()->isPointerTy() &&
+         V->getType()
+             ->getPointerElementType()
+             ->getPointerElementType()
+             ->isStructTy());
+  for (auto *U : V->users()) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                    dbgs() << "Analyzing user of " << *V << ": " << *U << "\n");
+    if (isa<Instruction>(U)) {
+      auto *Inst = dyn_cast<Instruction>(U);
+      if (Inst->getOpcode() == Instruction::Load) {
```

Comment:
Use lambda to share the near-identical code for the instruction case and operator case.